tests/test_main.py

```diff
@@ -190,3 +190,64 @@ def test_write_results_from_dummy_data_file(dummy_data_fname, results_fname, tmp
     results_file = tmp_path / results_fname
     main.write_results(dummy_data_file, results_file)
     assert results_file.read_text(encoding="utf-8") == "id\n1\n2\n"
+
+
+@pytest.mark.parametrize(
+    "sql_query",
+    [
+        # A complex query with a common table expression and a correlated
+        # subquery and column aliases to ensure that the parser
+        # correctly identifies the output columns
+        """
+        ;with CTE as (
+        SELECT
+            column1,
+            [column 2],
+            count(*) as column_3
+        FROM
+            table
+        GROUP BY
+            column1,
+            [column 2],
+        )
+        SELECT c.*,x.column_4
+        FROM CTE c
+        JOIN (
+            SELECT
+                column1,
+                column_4
+            FROM table2
+        ) x
+            ON c.column1 = x.column1
+        """,
+        # A multi-statement batch to ensure we create headers for the last
+        # statement in the batch
+        """
+        SELECT foo, bar, baz, qux
+        INTO #temptable;
+        SELECT
+            foo as [column1],
+            bar as [column 2],
+            baz as [column_3],
+            qux as [column_4]
+        FROM #temptable
+        """,
+    ],
+)
+def test_generate_column_headers(sql_query):
+    results = main.generate_column_headers(sql_query)
+    header = next(results)
+    assert list(header.keys()) == ["column1", "column 2", "column_3", "column_4"]
+    # we want only a header, no row data
+    assert not next(results, None)
+
+
+def test_generate_column_headers_fails_with_star():
+    # Ensure that a helpful error message is generated
+    # if a user tries to generate columns from a query
+    # that has `*` in the select list
+    with pytest.raises(
+        RuntimeError,
+        match=r"Headers can only be generated for queries with explicit column names",
```

Comment:
Honestly, I don't think we need to match. Simply testing for `RuntimeError` is enough.

Reply:
Explicit is better than implicit

Reply:
Do consider the value of this test, though. For more on the value of tests, take a quick look at "[Unit Testing Principles, Practices, and Patterns](https://learning.oreilly.com/library/view/unit-testing-principles/9781617296277/)". (We can access O'Reilly with our Oxford SSO.)

Reply:
The value of this test is that we get the right runtime error when the user input doesn't meet the requirements. AFAICT there are a number of reasons a RuntimeError might be thrown and so it is worthwhile making sure that the test passes because the correct error was thrown not because of some other spurious reason. If you think there is something detrimental to matching this error message could you explain it?

Reply:
I've read [Unit Testing Principles, Practices, and Patterns](https://learning.oreilly.com/library/view/unit-testing-principles/9781617296277/) a while ago - I'm not spotting the antipattern that I think you're alluding to here

Reply:
It's less an antipattern, more an outlook. The book emphasizes writing high-quality tests; that is, tests that are worth more than their maintenance costs. I highlighted this section from Chapter 1:

> It’s easy to create tests whose net value is close to zero or even is negative due to high maintenance costs. To enable sustainable project growth, you have to exclusively focus on high-quality tests—those are the only type of tests that are worth keeping in the test suite.

In this instance, I think testing for a `RuntimeError` is a high-quality test. We anticipate that `get_column_headers` will not always work; we signal this to the client by raising a `RuntimeError` (although as we have discussed, this may not be the most suitable type of error). However, I think testing for the message is not a high-quality test. If the client isn't a human, then the message doesn't matter. If the client is a human, then this is one of several potential messages. What makes this message the best of the potential messages? To put that another way, if we changed the message, then what would the test failure tell us that we didn't know?

Anyhow, I think we're getting into the weeds. I'm happy to resolve this conversation without a change 🙂